### DISCUSSION TITLE
Creating bindings to external JS code

### DISCUSSION THREAD

**Discussion by WhoAteDaCake** [mint-lang/mint#475] [Language Development]
Hi, I was just wondering if you had any ideas or plans about bindings with external javascript languages (e.g. d3)? I guess it would function similar to how Crystal can be bound to existing C libraries.

I've seen reasonml (another compiled to js language) use the following approach: 
https://github.com/af/bs-d3/blob/master/src/D3_Line.re

**Comment by gdotdesign** [comment]
Hi @WhoAteDaCake :wave: 

Mint doesn't have FFI like ReasonML but instead a simpler JavaScript inlining which you can read about on the guide: https://www.mint-lang.com/guide/reference/javascript-interop

It works with all sorts of JavaScript libraries like D3.

**Reply by WhoAteDaCake** [reply]
It can only decode primitive values, right? If I want to return a class instance created `d3` to re-use it later to read the value or something along this line, how would I type it?

**Reply by gdotdesign** [reply]
Here is a simple example of using D3 https://sandbox.mint-lang.com/sandboxes/Tf54Ptl0zxTCnw I hope it helps, let me know if it's not clear enough.

To better understand the type system check out this page: https://www.mint-lang.com/guide/reference/type-system

**Reply by WhoAteDaCake** [reply]
Oohhhh, that makes sense, thank you for this!

**Reply by frnco** [reply]
> Here is a simple example of using D3 https://sandbox.mint-lang.com/sandboxes/Tf54Ptl0zxTCnw I hope it helps, let me know if it's not clear enough.
> 
> To better understand the type system check out this page: https://www.mint-lang.com/guide/reference/type-system

This is pretty nice. One of the things I love about Mint is how concise it is when compared to stuff like Elm, and how flexible it is in most situations. The only issue that actually bothers me is how hard it is to find examples and documentation. The sandbox listing on the site could definitely use some way to search/filter, but even simply including links to examples like this one in relevant pages on the learn section would be VERY helpful.
